### PR TITLE
Pad the perm table to avoid array out of bound

### LINKED ISSUE #75224: Vxsort on Windows x64 accessing past the end of the buffer on some runtime tests
As part of the work to bring AddressSanitizer online, I've found a violation in vxsort in the GC. I've tried to repro-locally, but have not had success (I have a feeling that my machine has too much memory to repro, and when I tried setting the heap hard limit to something lower, I hit https://github.com/dotnet/runtime/issues/75214).

Here's the report from AddressSanitizer, and a link to the Helix work item's console log that hit the failure: https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-74623-merge-969884b56a1446e2ba/JIT.1/1/console.a2a09e5b.log?helixlogtype=result

<details>
<summary>AddressSanitizer report</summary>

```
      =================================================================
      ==3752==ERROR: AddressSanitizer: global-buffer-overflow on address 0x7ffb65a0f7f8 at pc 0x7ffb64ef99b7 bp 0x008a0517ad00 sp 0x008a0517ad00
      READ of size 16 at 0x7ffb65a0f7f8 thread T0
          #0 0x7ffb64ef99b6 in vxsort::vxsort<int,1,8,0>::align_vectorized D:\a\_work\1\s\src\coreclr\gc\vxsort\vxsort.h:587
          #1 0x7ffb64ef760f in vxsort::vxsort<int,1,8,0>::vectorized_partition<8> D:\a\_work\1\s\src\coreclr\gc\vxsort\vxsort.h:398
          #2 0x7ffb64efae45 in vxsort::vxsort<int,1,8,0>::sort D:\a\_work\1\s\src\coreclr\gc\vxsort\vxsort.h:275
          #3 0x7ffb64efb0cf in vxsort::vxsort<int,1,8,0>::sort D:\a\_work\1\s\src\coreclr\gc\vxsort\vxsort.h:630
          #4 0x7ffb64efb2c5 in vxsort::vxsort<__int64,1,8,3>::sort D:\a\_work\1\s\src\coreclr\gc\vxsort\vxsort.h:217
          #5 0x7ffb64efb741 in vxsort::vxsort<__int64,1,8,3>::sort D:\a\_work\1\s\src\coreclr\gc\vxsort\vxsort.h:630
          #6 0x7ffb64efa024 in do_vxsort_avx2 D:\a\_work\1\s\src\coreclr\gc\vxsort\do_vxsort_avx2.cpp:17
          #7 0x7ffb64eba6e2 in WKS::do_vxsort D:\a\_work\1\s\src\coreclr\gc\gc.cpp:9839
          #8 0x7ffb64ed4511 in WKS::gc_heap::plan_phase D:\a\_work\1\s\src\coreclr\gc\gc.cpp:28832
          #9 0x7ffb64ec03dc in WKS::gc_heap::gc1 D:\a\_work\1\s\src\coreclr\gc\gc.cpp:20981
          #10 0x7ffb64ec018e in WKS::gc_heap::garbage_collect D:\a\_work\1\s\src\coreclr\gc\gc.cpp:22747
          #11 0x7ffb64e9be22 in WKS::GCHeap::GarbageCollectGeneration D:\a\_work\1\s\src\coreclr\gc\gc.cpp:46936
          #12 0x7ffb64ee7b3a in WKS::gc_heap::trigger_gc_for_alloc D:\a\_work\1\s\src\coreclr\gc\gc.cpp:17691
          #13 0x7ffb64ee829d in WKS::gc_heap::try_allocate_more_space D:\a\_work\1\s\src\coreclr\gc\gc.cpp:17841
          #14 0x7ffb64e99cb0 in WKS::GCHeap::Alloc D:\a\_work\1\s\src\coreclr\gc\gc.cpp:45894
          #15 0x7ffb648515dc in Alloc D:\a\_work\1\s\src\coreclr\vm\gchelpers.cpp:226
          #16 0x7ffb64855817 in AllocateSzArray D:\a\_work\1\s\src\coreclr\vm\gchelpers.cpp:478
          #17 0x7ffb6490ec76 in JIT_NewArr1 D:\a\_work\1\s\src\coreclr\vm\jithelpers.cpp:2625
          #18 0x7ffb05f64707  (<unknown module>)
      
      0x7ffb65a0f800 is located 0 bytes to the right of global variable 'vxsort::perm_table_32' defined in 'machine_traits.avx2.cpp:28:27' (0x7ffb65a0f000) of size 2048
      SUMMARY: AddressSanitizer: global-buffer-overflow D:\a\_work\1\s\src\coreclr\gc\vxsort\vxsort.h:587 in vxsort::vxsort<int,1,8,0>::align_vectorized
      Shadow bytes around the buggy address:
        0x11d9e1ec1ea0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
        0x11d9e1ec1eb0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
        0x11d9e1ec1ec0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
        0x11d9e1ec1ed0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
        0x11d9e1ec1ee0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
      =>0x11d9e1ec1ef0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00[00]
        0x11d9e1ec1f00: f9 f9 f9 f9 f9 f9 f9 f9 f9 f9 f9 f9 f9 f9 f9 f9
        0x11d9e1ec1f10: f9 f9 f9 f9 f9 f9 f9 f9 f9 f9 f9 f9 f9 f9 f9 f9
        0x11d9e1ec1f20: f9 f9 f9 f9 f9 f9 f9 f9 f9 f9 f9 f9 f9 f9 f9 f9
        0x11d9e1ec1f30: f9 f9 f9 f9 f9 f9 f9 f9 f9 f9 f9 f9 f9 f9 f9 f9
        0x11d9e1ec1f40: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
      Shadow byte legend (one shadow byte represents 8 application bytes):
        Addressable:           00
        Partially addressable: 01 02 03 04 05 06 07 
        Heap left redzone:       fa
        Freed heap region:       fd
        Stack left redzone:      f1
        Stack mid redzone:       f2
        Stack right redzone:     f3
        Stack after return:      f5
        Stack use after scope:   f8
        Global redzone:          f9
        Global init order:       f6
        Poisoned by user:        f7
        Container overflow:      fc
        Array cookie:            ac
        Intra object redzone:    bb
        ASan internal:           fe
        Left alloca redzone:     ca
        Right alloca redzone:    cb
        Shadow gap:              cc
      ==3752==ABORTING
```

</details>

It looks like vxsort is reading a 16-byte vector when there's only 8 bytes left in the array. I have a feeling that vxsort handles this by chopping off the last 8 bytes if it realizes that it overran the buffer, but instead it should pad the global buffer so we aren't reading past the end of the buffer in the first place.

### PR DESCRIPTION
**Pull request by cshung**
Fixes #75224

### CODE REVIEW

File: src/coreclr/gc/vxsort/machine_traits.avx2.cpp [hunk 1]
```diff
@@ -23,6 +23,7 @@ alignas(128) const int8_t perm_table_64[128] = {
         2, 3, 0, 1, 4, 5, 6, 7,  // 0b1101 (13)
         0, 1, 2, 3, 4, 5, 6, 7,  // 0b1110 (14)
         0, 1, 2, 3, 4, 5, 6, 7,  // 0b1111 (15)
+        0, 0, 0, 0, 0, 0, 0, 0
 };
 
 alignas(2048) const int8_t perm_table_32[2048] = {
```
Comment (EgorBo) [inline src/coreclr/gc/vxsort/machine_traits.avx2.cpp:29]: I wonder if alignas(2048) is needed here, should not it be 32 instead?

**Comment by PeterSolMS** [comment]
Looks good to me.

Small nits: You may want to express the magic values 136 and 2056 as 128+8 and 2048 + 8 and add a comment explaining why there are 8 bytes of padding. This just for future maintenance...

**Review by PeterSolMS — APPROVED** [review]

File: src/coreclr/gc/vxsort/machine_traits.avx2.h [hunk 2]
```diff
@@ -24,8 +24,28 @@
...
+// into a single instruction that will not read more that 4/8 bytes..
+// vpmovsxb[dq] ymm0, dword [...], eliminating the 128-bit load completely and effectively
+// reading exactly 4/8 (depending if the instruction is vpmovsxbd or vpmovsxbq)
+// without generating an out of bounds read at all.
+// But, life is harsh, and we can't trust the compiler to do the right thing if it is not
+// contractual, hence this flustercuck
+#if !defined(__has_feature)
+#define __has_feature(a)    (0)
+#endif
+#if !__has_feature(address_sanitizer)
```
Comment (mangod9) [inline src/coreclr/gc/vxsort/machine_traits.avx2.h:39]: are there compilers where this isnt supported?

**Comment by jkotas** [comment]
Does this mean that this working around a bug in the address sanitizer? Is there an issue somewhere that tracks this bug so that we can remove the workaround once it is fixed?

File: src/coreclr/gc/vxsort/machine_traits.avx2.h [hunk 3]
```diff
@@ -24,8 +24,28 @@
...
-extern const int8_t perm_table_32[2048];
+
+// We might read the last 8 bytes into a 128-bit vector for the purpose of permutation
+// Most compilers actually fuse the pair of instructions: _mm256_cvtepi8_epiNN + _mm_loadu_si128
+// into a single instruction that will not read more that 4/8 bytes..
+// vpmovsxb[dq] ymm0, dword [...], eliminating the 128-bit load completely and effectively
+// reading exactly 4/8 (depending if the instruction is vpmovsxbd or vpmovsxbq)
+// without generating an out of bounds read at all.
+// But, life is harsh, and we can't trust the compiler to do the right thing if it is not
+// contractual, hence this flustercuck
```
Comment (jkotas) [inline src/coreclr/gc/vxsort/machine_traits.avx2.h:35]:
```suggestion
// contractual
```

File: src/coreclr/gc/vxsort/machine_traits.avx2.h [hunk 4]
```diff
@@ -24,8 +24,28 @@
...
-extern const int8_t perm_table_64[128];
-extern const int8_t perm_table_32[2048];
+
+// We might read the last 8 bytes into a 128-bit vector for the purpose of permutation
+// Most compilers actually fuse the pair of instructions: _mm256_cvtepi8_epiNN + _mm_loadu_si128
+// into a single instruction that will not read more that 4/8 bytes..
+// vpmovsxb[dq] ymm0, dword [...], eliminating the 128-bit load completely and effectively
+// reading exactly 4/8 (depending if the instruction is vpmovsxbd or vpmovsxbq)
+// without generating an out of bounds read at all.
+// But, life is harsh, and we can't trust the compiler to do the right thing if it is not
```
Comment (jkotas) [inline src/coreclr/gc/vxsort/machine_traits.avx2.h:34]: If I am reading this correctly, we are trusting the compiler to do the "right thing" when it does not have `address_sanitizer` feature. Why is that a good idea?

Comment (cshung) [reply src/coreclr/gc/vxsort/machine_traits.avx2.h:34]: The comment was copied from upstream [here](https://github.com/damageboy/vxsort-cpp/blob/096e967cd6a3be2b18d4baf1af81893d97fad7dd/vxsort/machine_traits.avx2.h#L23)

and the pattern of including the padding bytes conditionally was copied [here](https://github.com/dotnet/runtime/blob/4cf1383c8458945b7eb27ae5f57338c10ed25d54/src/coreclr/gc/vxsort/smallsort/avx2_load_mask_tables.cpp#L16)

By the time I was copying them, I didn't notice they contradict each other.

Perhaps we should remove the `#if !__has_feature(address_sanitizer)` to stay on the safe side.

File: src/coreclr/gc/vxsort/machine_traits.avx2.cpp [hunk 5]
```diff
@@ -23,9 +23,10 @@ alignas(128) const int8_t perm_table_64[128] = {
         2, 3, 0, 1, 4, 5, 6, 7,  // 0b1101 (13)
         0, 1, 2, 3, 4, 5, 6, 7,  // 0b1110 (14)
         0, 1, 2, 3, 4, 5, 6, 7,  // 0b1111 (15)
+        0xCC, 0xCC, 0xCC, 0xCC, 0xCC, 0xCC, 0xCC, 0xCC, // Garbage to make ASAN happy
```
[the remark this answers is not in the document]
Comment (jkotas) [reply src/coreclr/gc/vxsort/machine_traits.avx2.cpp:26]: (like what you have done in `perm_table_32` table)

**Review by jkotas — APPROVED** [review]

**Pull request merged**